provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/Searcher.java

```diff
@@ -344,4 +345,68 @@ public static List<PublicId> getAllowedResultsFromResultConformance(NavigationCa
         return allowedResultsList;
     }
 
+    /**
+     * Returns a list of PublicIds for Descendants given an ancestor and list of ancestors to constrain by
+     *
+     * @param ancestorId            ID of the primary ancestor to find Descendants of
+     * @param constraintAncestorIds List of IDs for descendants to constrain by (includes itself)
+     * @return List of PublicIds for a given ancestor constrained by other ancestors
+     */
+    public static List<PublicId> getDescendantsWithConstraint(PublicId ancestorId, List<PublicId> constraintAncestorIds) {
+        List<PublicId> descendantsWithConstraints = new ArrayList<>();
+        if (PrimitiveData.get().hasPublicId(ancestorId)) {
+            EntityService.get().getEntity(ancestorId.asUuidArray()).ifPresent(ancestorEntity -> {
+                if (ancestorEntity instanceof Concept) {
+                    List<PublicId> constraintDescendants = new ArrayList<>();
+                    descendantsWithConstraints.addAll(descendantsOf(ancestorId));
+                    constraintAncestorIds.forEach(constraintId -> {
+                        if (PrimitiveData.get().hasPublicId(constraintId)) {
+                            EntityService.get().getEntity(constraintId.asUuidArray()).ifPresent(constraintEntity -> {
+                                if (constraintEntity instanceof Concept) {
+                                    constraintDescendants.add(constraintId);
+                                    constraintDescendants.addAll(descendantsOf(constraintId));
+                                }
+                            });
+                        }
+                    });
+                    descendantsWithConstraints.removeAll(constraintDescendants);
+                }
+            });
+        }
+        return descendantsWithConstraints;
+    }
+
+    /**
+     * Returns a list of PublicIds for Descendants given an ancestor and list of ancestors to constrain by
+     *
+     * @param navCalc               NavigationCalculator for determining latest version
+     * @param ancestorId            ID of the primary ancestor to find Descendants of
+     * @param constraintAncestorIds List of IDs for descendants to constrain by (includes itself)
+     * @return List of PublicIds for a given ancestor constrained by other ancestors
+     */
+    public static List<PublicId> getDescendantsWithConstraint(NavigationCalculator navCalc, PublicId ancestorId, List<PublicId> constraintAncestorIds) {
```

Comment:
These two methods are almost identical except for the call to "decendantsOf". You could extract this call to a functional interface/lambda. Something like the following:

create a third private helper method that takes a functional interface and contains the bulk of the code:

`    private static List<PublicId> getDescendantsWithConstraint(PublicId ancestorId, List<PublicId> constraintAncestorIds, BiFunction<NavigationCalculator, PublicId, List<PublicId>> func) {...}`

In body of the method replace "decendantsOf(...)" with "func.apply(...)"

Then each of the public methods becomes a one liner, passing in a lambda with their specific call to "decendantsOf". Something like:

```
    public static List<PublicId> getDescendantsWithConstraint(NavigationCalculator navCalc, PublicId ancestorId, List<PublicId> constraintAncestorIds) {
        return getDescendantsWithConstraint(ancestorId, constraintAncestorIds, (nCalc, anc) -> descendantsOf(nCalc, anc));
    }
```